### DISCUSSION TITLE
$watch disable/deactivate

### DISCUSSION THREAD

**Discussion by sunsande** [alpinejs/alpine#3273] [1. Help] [answered]
The question in short: how to disable/deactivate  a `$watch` set on some model object/property?

Motivation: I have a model object with many fields (binded to `<input>` controls) and I want to be able to check if it was change by the user (on dirty). It is not the perfect check (if the user enters a letter and then deletes it, there is no actual change, but it would still count as dirty in my situation). The object is loaded with fetch and directly after that a `$watch` is set on the whole object with callback setting a flag `dirty = true`. 

It works perfectly, but it continues to get fired after the first change was made by the user and obviously this is not needed. In my application it is not a problem as the callback does only `dirty = true`, but it would be nice to be able to control the pointless fire of the callback when it is not needed anymore.

Is there maybe another way to achieve that functionality (without `watch`)? (I am aware there is the `effect` function but it would have the same problem, or am I wrong?)

**Comment by ekwoka** [comment, marked as answer]
Alpine.effect and Alpine.release can do this just fine.

**Reply by sunsande** [reply]
@ekwoka thanks a lot for the reply! Could you please give some explanation or some example for the exact syntax.
I did not find anything on the `Alpine.release` in the documentation and could not find it out from a short 15 minutes look through the source code.
The function `Alpine.release` is there: if I run it without parameters it gives an error. I called it with `Alpine.release('someProperty')` and even with `Alpine.release(theDataObject)` without success - it runs, but nothing happens - the `Alpine.effect` callback continues to run after such calls.
P.S. : does this function works on $watch also?

**Reply by ekwoka** [reply]
Sorry.

Here's an example

```js
const myEffect = Alpine.effect(()=>console.log(this.user))

// Will console log whenever user changes

// later
Alpine.release(myEffect)

// Will no longer console log when use changes

```

**Reply by sunsande** [reply]
Thanks a lot!
And if it is not too much to ask, is there a way to get a reference to the $watch effect (in order to be able to delete/release it)? (I saw in the source code that it is used , but deleted afterwards from the element's effects list)

**Reply by ekwoka** [reply]
Watches are removed when the component itself unmounts. There is no other "proper", way to release a watch effect. Only internal "hacky", ways.

**Reply by sunsande** [reply]
> ```js
> const myEffect = Alpine.effect(()=>console.log(this.user))
> ```
actually this does not work (no console.log happens on user change), when `this.user` is a complex object (for example when this.user = { name: 'x' } - and this works with `$watch` perfectly)

**Reply by SimoTod** [reply]
`console.log(JSON.stringify(this.user));` Alpine needs to touch every single variable to track them correctly.

As a reference, this is how $watch is implemented: https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/magics/%24watch.js

**Reply by sunsande** [reply]
@SimoTod  Thanks a lot! Putting the same trick (`JSON.stringify`) into the Alpine.effect made it work on the complex object.
